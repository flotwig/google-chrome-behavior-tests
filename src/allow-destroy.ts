import net from 'net'

export type Destroyable<T> = T & {
  destroy: (cb: Function) => void
}

/**
 * `allowDestroy` adds a `destroy` method to a `net.Server`. `destroy(cb)`
 * will kill all open connections and call `cb` when the server is closed.
 *
 * Note: `server-destroy` NPM package cannot be used - it does not track
 * `secureConnection` events.
 */
export function allowDestroy <T extends net.Server>(server: T): Destroyable<T> {
  let connections: net.Socket[] = []

  function trackConn (conn: net.Socket) {
    connections.push(conn)

    conn.on('close', () => {
      connections = connections.filter((connection) => connection !== conn)
    })
  }

  server.on('connection', trackConn)
  server.on('secureConnection', trackConn)

  // @ts-ignore Property 'destroy' does not exist on type 'Server'.
  server.destroy = function (cb) {
    server.close(cb)
    connections.map((connection) => connection.destroy())
  }

  return server as Destroyable<T>
}
