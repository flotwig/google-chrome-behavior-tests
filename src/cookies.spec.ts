import _ from 'lodash'
import { allowDestroy, Destroyable } from './allow-destroy'
import express from 'express'
import morgan from 'morgan'
import os from 'os'
import path from 'path'
import puppeteer from 'puppeteer'
import { promises } from 'fs'
import { Server } from 'http'

const fs = promises

const PORT = 12344

const getNewTempDir = () => fs.mkdtemp(path.join(os.tmpdir(), 'chrome-test-profile-'))

const launchBrowser = async () =>
  puppeteer.launch({
    userDataDir: await getNewTempDir(),
    args: [
      `--proxy-server=http://localhost:${PORT}`,
      `--no-sandbox`
    ]
  })

describe('Cookies', () => {
  let server: Destroyable<Server>
  let app: express.Application

  ([
    ['original domain', 8],
    ['different domain', 7]
  ]).forEach(([ending, N]) => {
    it(`sets cookies on a long list of redirects, ending with ${ending}`, async () => {
      const browser = await launchBrowser()
      const page = await browser.newPage()

      app = express()
      app.use(morgan('common'))

      const cascadingHandler: express.Handler = function(req, res) {
        var a, b, n;
        n = Number(req.query.n);
        // alternates between domains
        a = req.query.a;
        b = req.query.b;
        res.header("Set-Cookie", [`namefoo${n}=valfoo${n}`, `namebar${n}=valbar${n}`]);
        if (n > 0) {
          res.redirect(`${a}/setCascadingCookies?n=${n - 1}&a=${b}&b=${a}`);
        }
        return res.send("<html>finished setting cookies</html>");
      }

      app.get('*', cascadingHandler);

      server = allowDestroy(app.listen(PORT))

      await page.goto(`http://site1.foo.com/setCascadingCookies?n=${N}&a=http://site2.bar.net&b=http://site1.foo.com`)

      let cookies = await page.cookies('http://site1.foo.com', 'http://site2.bar.net')
      cookies = _.reverse(_.sortBy(cookies, _.property('name')))

      expect(cookies).toMatchSnapshot()

      await browser.close()
      await new Promise((r) => server.destroy(r))
    })
  })
})
