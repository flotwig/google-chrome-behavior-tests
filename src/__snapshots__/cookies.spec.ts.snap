// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cookies multiple Set-Cookie with one cookie each sets cookies on a long list of redirects, ending with different domain 1`] = `
Array [
  Object {
    "domain": "site1.foo.com",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo7",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valfoo7",
  },
  Object {
    "domain": "site2.bar.net",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo6",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valfoo6",
  },
  Object {
    "domain": "site1.foo.com",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo5",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valfoo5",
  },
  Object {
    "domain": "site2.bar.net",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo4",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valfoo4",
  },
  Object {
    "domain": "site1.foo.com",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo3",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valfoo3",
  },
  Object {
    "domain": "site2.bar.net",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo2",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valfoo2",
  },
  Object {
    "domain": "site1.foo.com",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo1",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valfoo1",
  },
  Object {
    "domain": "site2.bar.net",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo0",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valfoo0",
  },
  Object {
    "domain": "site1.foo.com",
    "expires": -1,
    "httpOnly": false,
    "name": "namebar7",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valbar7",
  },
  Object {
    "domain": "site2.bar.net",
    "expires": -1,
    "httpOnly": false,
    "name": "namebar6",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valbar6",
  },
  Object {
    "domain": "site1.foo.com",
    "expires": -1,
    "httpOnly": false,
    "name": "namebar5",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valbar5",
  },
  Object {
    "domain": "site2.bar.net",
    "expires": -1,
    "httpOnly": false,
    "name": "namebar4",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valbar4",
  },
  Object {
    "domain": "site1.foo.com",
    "expires": -1,
    "httpOnly": false,
    "name": "namebar3",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valbar3",
  },
  Object {
    "domain": "site2.bar.net",
    "expires": -1,
    "httpOnly": false,
    "name": "namebar2",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valbar2",
  },
  Object {
    "domain": "site1.foo.com",
    "expires": -1,
    "httpOnly": false,
    "name": "namebar1",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valbar1",
  },
  Object {
    "domain": "site2.bar.net",
    "expires": -1,
    "httpOnly": false,
    "name": "namebar0",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valbar0",
  },
]
`;

exports[`Cookies multiple Set-Cookie with one cookie each sets cookies on a long list of redirects, ending with original domain 1`] = `
Array [
  Object {
    "domain": "site1.foo.com",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo8",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valfoo8",
  },
  Object {
    "domain": "site2.bar.net",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo7",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valfoo7",
  },
  Object {
    "domain": "site1.foo.com",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo6",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valfoo6",
  },
  Object {
    "domain": "site2.bar.net",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo5",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valfoo5",
  },
  Object {
    "domain": "site1.foo.com",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo4",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valfoo4",
  },
  Object {
    "domain": "site2.bar.net",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo3",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valfoo3",
  },
  Object {
    "domain": "site1.foo.com",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo2",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valfoo2",
  },
  Object {
    "domain": "site2.bar.net",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo1",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valfoo1",
  },
  Object {
    "domain": "site1.foo.com",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo0",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valfoo0",
  },
  Object {
    "domain": "site1.foo.com",
    "expires": -1,
    "httpOnly": false,
    "name": "namebar8",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valbar8",
  },
  Object {
    "domain": "site2.bar.net",
    "expires": -1,
    "httpOnly": false,
    "name": "namebar7",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valbar7",
  },
  Object {
    "domain": "site1.foo.com",
    "expires": -1,
    "httpOnly": false,
    "name": "namebar6",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valbar6",
  },
  Object {
    "domain": "site2.bar.net",
    "expires": -1,
    "httpOnly": false,
    "name": "namebar5",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valbar5",
  },
  Object {
    "domain": "site1.foo.com",
    "expires": -1,
    "httpOnly": false,
    "name": "namebar4",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valbar4",
  },
  Object {
    "domain": "site2.bar.net",
    "expires": -1,
    "httpOnly": false,
    "name": "namebar3",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valbar3",
  },
  Object {
    "domain": "site1.foo.com",
    "expires": -1,
    "httpOnly": false,
    "name": "namebar2",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valbar2",
  },
  Object {
    "domain": "site2.bar.net",
    "expires": -1,
    "httpOnly": false,
    "name": "namebar1",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valbar1",
  },
  Object {
    "domain": "site1.foo.com",
    "expires": -1,
    "httpOnly": false,
    "name": "namebar0",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valbar0",
  },
]
`;

exports[`Cookies single Set-Cookie with comma-separated multiple cookies sets cookies on a long list of redirects, ending with different domain 1`] = `
Array [
  Object {
    "domain": "site1.foo.com",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo7",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 33,
    "value": "valfoo7, namebar7=valbar7",
  },
  Object {
    "domain": "site2.bar.net",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo6",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 33,
    "value": "valfoo6, namebar6=valbar6",
  },
  Object {
    "domain": "site1.foo.com",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo5",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 33,
    "value": "valfoo5, namebar5=valbar5",
  },
  Object {
    "domain": "site2.bar.net",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo4",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 33,
    "value": "valfoo4, namebar4=valbar4",
  },
  Object {
    "domain": "site1.foo.com",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo3",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 33,
    "value": "valfoo3, namebar3=valbar3",
  },
  Object {
    "domain": "site2.bar.net",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo2",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 33,
    "value": "valfoo2, namebar2=valbar2",
  },
  Object {
    "domain": "site1.foo.com",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo1",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 33,
    "value": "valfoo1, namebar1=valbar1",
  },
  Object {
    "domain": "site2.bar.net",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo0",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 33,
    "value": "valfoo0, namebar0=valbar0",
  },
]
`;

exports[`Cookies single Set-Cookie with comma-separated multiple cookies sets cookies on a long list of redirects, ending with original domain 1`] = `
Array [
  Object {
    "domain": "site1.foo.com",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo8",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 33,
    "value": "valfoo8, namebar8=valbar8",
  },
  Object {
    "domain": "site2.bar.net",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo7",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 33,
    "value": "valfoo7, namebar7=valbar7",
  },
  Object {
    "domain": "site1.foo.com",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo6",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 33,
    "value": "valfoo6, namebar6=valbar6",
  },
  Object {
    "domain": "site2.bar.net",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo5",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 33,
    "value": "valfoo5, namebar5=valbar5",
  },
  Object {
    "domain": "site1.foo.com",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo4",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 33,
    "value": "valfoo4, namebar4=valbar4",
  },
  Object {
    "domain": "site2.bar.net",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo3",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 33,
    "value": "valfoo3, namebar3=valbar3",
  },
  Object {
    "domain": "site1.foo.com",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo2",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 33,
    "value": "valfoo2, namebar2=valbar2",
  },
  Object {
    "domain": "site2.bar.net",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo1",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 33,
    "value": "valfoo1, namebar1=valbar1",
  },
  Object {
    "domain": "site1.foo.com",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo0",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 33,
    "value": "valfoo0, namebar0=valbar0",
  },
]
`;

exports[`Cookies single Set-Cookie with semicolon-separated multiple cookies sets cookies on a long list of redirects, ending with different domain 1`] = `
Array [
  Object {
    "domain": "site1.foo.com",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo7",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valfoo7",
  },
  Object {
    "domain": "site2.bar.net",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo6",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valfoo6",
  },
  Object {
    "domain": "site1.foo.com",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo5",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valfoo5",
  },
  Object {
    "domain": "site2.bar.net",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo4",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valfoo4",
  },
  Object {
    "domain": "site1.foo.com",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo3",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valfoo3",
  },
  Object {
    "domain": "site2.bar.net",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo2",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valfoo2",
  },
  Object {
    "domain": "site1.foo.com",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo1",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valfoo1",
  },
  Object {
    "domain": "site2.bar.net",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo0",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valfoo0",
  },
]
`;

exports[`Cookies single Set-Cookie with semicolon-separated multiple cookies sets cookies on a long list of redirects, ending with original domain 1`] = `
Array [
  Object {
    "domain": "site1.foo.com",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo8",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valfoo8",
  },
  Object {
    "domain": "site2.bar.net",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo7",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valfoo7",
  },
  Object {
    "domain": "site1.foo.com",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo6",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valfoo6",
  },
  Object {
    "domain": "site2.bar.net",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo5",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valfoo5",
  },
  Object {
    "domain": "site1.foo.com",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo4",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valfoo4",
  },
  Object {
    "domain": "site2.bar.net",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo3",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valfoo3",
  },
  Object {
    "domain": "site1.foo.com",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo2",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valfoo2",
  },
  Object {
    "domain": "site2.bar.net",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo1",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valfoo1",
  },
  Object {
    "domain": "site1.foo.com",
    "expires": -1,
    "httpOnly": false,
    "name": "namefoo0",
    "path": "/",
    "secure": false,
    "session": true,
    "size": 15,
    "value": "valfoo0",
  },
]
`;
